contribs/gmf/src/directives/timeslider.js

```diff
@@ -0,0 +1,170 @@
+        } else {
+          //element is hidden. Do not init tooltip now
+          $('body').on('shown.bs.popover', function(jqEvt) {
+            //Element contained in a popover ?
+            if ($('.popover[role="tooltip"]').find(element).length) {
+              element.find('span.ui-slider-handle').each(initToolTip_);
+            }
+          });
+        }
+
+        element.find('.ui-slider-handle').on('hide.bs.tooltip', function(jqEvt) {
+          if (element.find(this).length > 0) {
+            jqEvt.preventDefault();
+          }
+        });
+      }
+
+      function updateTooltip_(e, slider) {
+        var handle = $(slider.handle);
+        var left  = handle.position().left;
+        var tooltip = $('#' + handle.attr('aria-describedby'));
+        var dateText  = $filter('date')(new Date(slider.value), dateFormat);
+        handle.attr('data-original-title', dateText).tooltip('fixTitle');
+        tooltip.css('left', left - tooltip.width() / 2);
+        tooltip.find('.tooltip-inner').text(dateText);
+      }
+
+      function initToolTip_(i) {
+        var el = $(this);
+        if (!el.attr('aria-describedby')) {
+          var date = ctrl.dates[i] || ctrl.dates;
+          var dateText = $filter('date')(new Date(date), dateFormat);
+
+          el.tooltip({
+            'selector': 'span.ui-slider-handle',
+            'trigger': 'manual',
+            'placement' : i === 0 ? 'bottom' : 'top',
+            'title' : dateText
+          })
+          .tooltip('show');
+        }
+      }
+
+      function onSliderReleased_(e, sliderUi) {
+        var sDate, eDate;
+        updateTooltip_(e, sliderUi);
+        if (sliderUi.values) {
+          sDate = new Date(sliderUi.values[0]);
+          eDate = new Date(sliderUi.values[1]);
+        } else {
+          sDate = new Date(sliderUi.value);
+        }
+
+        ctrl.onDateSelected({
+          time : {
+            start : sDate,
+            end : eDate
+          }
+        });
+      }
+    }
+  };
+};
+
+gmf.module.directive('gmfTimeSlider', gmf.TimeSlider);
+
+
+/**
+ * TimeSliderController - directive conttroller
+ * @param {!angular.Scope} $scope Angular scope.
+ * @constructor
+ * @export
+ * @ngInject
+ * @ngdoc controller
+ * @ngname gmfTimeSliderController
+ */
+gmf.TimeSliderController = function($scope) {
+
+
+  /**
+   * @function
+   * @export
+   */
+  this.onDateSelected;
+
+
+  /**
+   * @type {Object}
+   * @export
+   */
+  this.time;
+
+
+  var min = new Date(this.time.minValue).getTime();
+  var max = new Date(this.time.maxValue).getTime();
+  var step = 1000 * 3600 * 24; //day resolution
+
+  /**
+   * @type {Object}
+   * @export
+   */
+  this.sliderOptions = {
+    'range' : this.time.mode === 'range',
+    'min' : min,
+    'max' : max,
+    'step' : step
+  };
+
+  /**
+   * @type {Array|number}
+   * @export
+   */
+  this.dates = this.time.mode === 'range' && [min, max] || Math.ceil(min + (max - min) / 2);
```

Comment:
That's hard to read.

Can you write it like this:

````
this.dates = this.time.mode === 'range' ? [min, max] : Math.ceil(min + (max - min) / 2);
````
Or even with if/else...

Let the compilator transforms it into something quick to read for machines.